src/macros.rs

```diff
@@ -1,43 +1,35 @@
 /// `frame!` can be used to create new local variables that play nicely with the garbage collector
 #[macro_export]
 macro_rules! frame {
-    (($($param:ident),*) $code:block) => {
-       {
-            #[allow(unused_unsafe)]
-            let caml_frame = unsafe { $crate::sys::local_roots() };
-            #[allow(unused_mut)]
-            $(let mut $param = $crate::Value($crate::sys::UNIT);)*
-            #[allow(unused_unsafe)]
-            $crate::sys::caml_param!($($param.0),*);
-            #[allow(unused_mut)]
-            let mut res = || { $code };
-            let res = res();
-
-            #[allow(unused_unsafe)]
-            unsafe { $crate::sys::set_local_roots(caml_frame) };
-            res
-        }
-    }
-}
+    ($gc:ident: ($($param:ident),*) $code:block) => {
+        {
+            struct __Values  {
+               $($param: $crate::Value),*
+            }
 
-/// `body!` is needed to help the OCaml runtime to manage garbage collection, it should
-/// be used to wrap the body of each function exported to OCaml.
-///
-/// ```rust
-/// #[no_mangle]
-/// pub extern "C" fn example(a: ocaml::Value, b: ocaml::Value) -> ocaml::Value {
-///     ocaml::body!((a, b) {
-///         let a = a.int_val();
-///         let b = b.int_val();
-///         ocaml::Value::int(a + b)
-///     })
-/// }
-/// ```
-#[macro_export]
-#[cfg(feature = "no-std")]
-macro_rules! body {
-    ($(($($param:expr),*))? $code:block) => {
-        $crate::sys::caml_body!($(($($param.0),*))? $code);
+            $crate::interop::ocaml_frame!($gc, ($($param),*), {
```

Comment:
note that `ocaml_frame!` only makes sense when there is at least one root, when there are no roots there is no need to open a frame. So this macro should use `+` instead of `*`.